### ISSUE TITLE
Cant get Parity.personalUnlockAccount(-) to work

### ISSUE THREAD

**Issue by MaxBinnewies** [web3j/web3j#8]

**Comment by MaxBinnewies** [comment]
Hi Conor,

Im afraid I have to bug you again.
I am trying to send transactions that modify the blockchain now. I have my own Parity node running, my account is imported(and Parity shows it and I can send ether from it). I also know the acount's public and private key aswell as the password. I thought I try Parity.personalUnlockAccount(...), but I cant get it to work, even with the code from the integration test. 

(1) This is what im trying:
	```
public static void main(String[] args) throws Exception {
		Parity parity = Parity.build(new HttpService());
		Web3ClientVersion web3ClientVersion = parity.web3ClientVersion().sendAsync().get();
		String clientVersion = web3ClientVersion.getWeb3ClientVersion();
		System.out.println(clientVersion);
		NewAccountIdentifier newAcc = parity.personalNewAccount("myPassword").send();
		PersonalListAccounts personalListAccounts = parity.personalListAccounts().send();
		PersonalUnlockAccount personalUnlockAccount = parity.personalUnlockAccount(newAcc.getAccountId(), "myPassword")
				.send();
		if (personalUnlockAccount.accountUnlocked()) {
			System.out.println("success");
		}
		System.out.println("end");
}
```
It nullpointers on the call of `"accountUnlocked(). This is what it shows for client version: "Parity//v1.3.9-beta-e9987c4-20161021/x86_64-windows-msvc/rustc1.12.0". personalListAccounts actually contains all the accounts, so web3j seems to have access to them. And personalNewAccount() also creates a new account file. But personalUnlockAccount() delivers an ResponseError with the code -32602 and the message "Invalid params". result is null. I have also tried it with my actual Ethereum-account, but got the same result. 

Do you know what I am doing wrong? Is it maybe a problem with the parity-version?

(2) If I get the account unlocked, how can I use that to send transactions that modify the state of the blockchain using the web3j-Contract Wrappers? If I pass the Parity object to the contract wrapper will that be enough to send transactions that use gas?

Thank you for your help,
Max

**Comment by conor10** [comment]
Hi Max,

Parity requires that account addresses are prefixed with "0x..." on the unlock account call - please can you confirm that this is the case in the JSON-RPC unlock request being sent?

I'm currently updating the contract wrappers to support transactions. I'm hoping to get that out this week.

The initial implementation will use offline transaction signing. But I can potentially add support for unlocking accounts too.

It's also worth being aware that Gav Wood is not too keen on the unlockAccount function, see https://github.com/ethcore/parity/issues/1215 for more info.

**Comment by MaxBinnewies** [comment]
Hi Conor,

thanks for the quick response!
I am not quite sure how to access the JSON-RPC to be honest.
But if I isolate the request object that is created by personalUnlockAccount() and debug into it with eclipse, this is the param field: 
[0xa4d79a29ccaeb0e8bcaf09b12010a665f821f191, myPassword]

Im not fixed on the unlockAccount function, I just thought this is the easiest way to go. What would you recommend instead?
Basically I am trying to call methods on my contract that modify the state of the blockchain. I have an account with Ether for which I know public key/private key, address and password. I could use my own parity node or infura. So, I am very flexible at the moment, just trying to get it to work.

Max

**Comment by conor10** [comment]
The easiest way to get it to work is to use the offline transaction signing - then you can work with Infura, and not have to worry about unlocking accounts & installing wallets with nodes. To extract your public/private key pair from your Ethereum wallet you can do this via myetherwallet.com (assuming you're using testnet). I'm working on getting keyfile support into web3j.

If you want to debug the JSON-RPC calls going to & from your Ethereum client, set the following system properties in your main class:

```
System.setProperty("org.apache.commons.logging.Log","org.apache.commons.logging.impl.SimpleLog");
System.setProperty("org.apache.commons.logging.simplelog.showdatetime", "true");
System.setProperty("org.apache.commons.logging.simplelog.log.org.apache.http.wire", "DEBUG");
```

**Comment by MaxBinnewies** [comment]
Hi Conor,

The JSON-RPC call is:
`2016/11/03 14:55:01:293 MEZ [DEBUG] wire - http-outgoing-0 >> "{"jsonrpc":"2.0","method":"personal_unlockAccount","params":["0xd3681e13da01b68eb6eb7c507cd96c324aacc130","myPassword"],"id":1}"`

However, it should look like this:
`'{"jsonrpc":"2.0","method":"personal_unlockAccount","params":["0xfc390d8a8ddb591b010fda52f4db4945742c3809","hunter2",null],"id": 22}'`

I believe the time parameter is not optional. JsonRpc2_0Parity.personalUnlockAccount(String accountId, String password, BigInteger duration) drops that parameter if its null in line 81 (0.6.0 I haven't updated yet), which is always the case if personalUnlockAccount(String accountId, String password) is called.

`Request<?, PersonalUnlockAccount> req = parity.personalUnlockAccount(newAcc.getAccountId(), "myPassword", 	new BigInteger("1"));
		req.getParams().set(2, null);`
seems to be working. At least .accountUnlocked() returns true. 
I will try to send transactions that cost gas this way now. Hope it works

Thanks for your help,
Max

**Comment by MaxBinnewies** [comment]
Ok, I just updated to 0.7.0.
Now I know what you mean by integrating Transactions into Contracts. This is awesome!!

There just seems to some problem in Credentials that I dont quite understand. Let me try to explain.

This is my public key:
0xa02b68609574223a5cd01db35ab337f66f6f3e1fc8e050b3ac7892e76a544f59c967abdaa9746363bdead18c24b512f82e537f59735e68cb7e79e948e5f25dd4

When I try to send a transaction, the Contracts class tries to calculate the next nonce by sending this JSON-RPC:
"{"jsonrpc":"2.0","method":"eth_getTransactionCount","params":["1d0fee96aa9750f87894b034d25e17edca8b76a3","latest"],"id":1}"
i remember you saying something about Parity always needing a "0x" at the beginning of the address. 
When I debug in Contract.executeTransaction(Function function), credentials.address is "1d0fee...", so when I add "0x" at the beginning while debugging, the method can find the correct nonce and the Transaction actually goes through and can modify the blockchain(yeay!)

How can I can fix this? I hope, I am not too annoying with all my little problems, your help is greatly appreciated! :-)

Also, I would like to recommend adding the JSON-RPC logging configuration, that you send me earlier in this issue, to the general web3j-documentation. That really helps A LOT while troubleshooting. I find it a lot easier to understand whats going on with this log.

Thank you,
Max

**Comment by conor10** [comment]
It's certainly not annoying - it's important that the library is easy to use for all users.

I've just put out a new release (1.0.1) which should address both issues you've raised - the Credentials class was not storing the address with a hex prefix, and I'm now sending 'null' for the duration field on account unlock requests.

I've added details of the JSON-RPC logging configuration to the troubleshooting section of the docs https://web3j.github.io/web3j/trouble.html#i-want-to-see-details-of-the-json-rpc-requests-and-responses.

Also, as of the 1.0 release web3j now provides support for working with wallet files if you want to use them instead.

Please reopen this ticket if you are still having any of the above issues.
